homeassistant/components/logbook/websocket_api.py

```diff
@@ -0,0 +1,304 @@
+) -> None:
+    """Handle logbook get events websocket command."""
+    start_time_str = msg["start_time"]
+    utc_now = dt_util.utcnow()
+
+    if start_time := dt_util.parse_datetime(start_time_str):
+        start_time = dt_util.as_utc(start_time)
+
+    if not start_time or start_time > utc_now:
+        connection.send_error(msg["id"], "invalid_start_time", "Invalid start_time")
+        return
+
+    device_ids = msg.get("device_ids")
+    entity_ids = msg.get("entity_ids")
+    if entity_ids:
+        entity_ids = async_filter_entities(hass, entity_ids)
+    event_types = async_determine_event_types(hass, entity_ids, device_ids)
+
+    event_processor = EventProcessor(
+        hass,
+        event_types,
+        entity_ids,
+        device_ids,
+        None,
+        timestamp=True,
+        include_entity_name=False,
+    )
+
+    stream_queue: asyncio.Queue[Event] = asyncio.Queue(MAX_PENDING_LOGBOOK_EVENTS)
+    subscriptions: list[CALLBACK_TYPE] = []
+    setup_complete_future: asyncio.Future[dt] = asyncio.Future()
+    task = asyncio.create_task(
+        _async_events_consumer(
+            setup_complete_future,
+            connection,
+            msg["id"],
+            stream_queue,
+            event_processor,
+        )
+    )
+
+    def _unsub() -> None:
+        """Unsubscribe from all events."""
+        for subscription in subscriptions:
+            subscription()
+        subscriptions.clear()
+        if task:
+            task.cancel()
+
+    @callback
+    def _queue_or_cancel(event: Event) -> None:
+        """Queue an event to be processed or cancel."""
+        try:
+            stream_queue.put_nowait(event)
+        except asyncio.QueueFull:
+            _LOGGER.debug(
+                "Client exceeded max pending messages of %s (likely disconnected without unsubscribe)",
+                MAX_PENDING_LOGBOOK_EVENTS,
+            )
+            _unsub()
+
+    async_subscribe_events(
+        hass, subscriptions, _queue_or_cancel, event_types, entity_ids, device_ids
+    )
+    subscriptions_setup_complete_time = dt_util.utcnow()
+    connection.subscriptions[msg["id"]] = _unsub
+    connection.send_result(msg["id"])
+
+    # Fetch everything from history
+    message, last_event_time = await _async_get_ws_formatted_events(
+        hass,
+        msg["id"],
+        start_time,
+        subscriptions_setup_complete_time,
+        messages.event_message,
+        event_processor,
+    )
+    # If there is not last_time there are not historical
+    # results, but we still send an empty message so
+    # consumers of the api know their request was
+    # answered but there were no results
+    connection.send_message(message)
+
+    if commit_interval := get_instance(hass).commit_interval:
+        # Fetch any events from the database that have
+        # not been committed since the original fetch
+        # so we can switch over to using the subscriptions
+        await asyncio.sleep(commit_interval)
```

Comment:
Could this be so long that it causes the subscriptions to fill up the queue ?

Reply:
It is definitely possible if they have a very busy system and a commit interval.

Since the option was added:

- We now only commit if there is actual work to do
- The amount of writes have been reduced significantly (for non-busy systems it is two orders of magnitude or better) 
- The sqlite database settings have been optimized to be much more friendly to sd cards 
- We don't close the sqlite connection anymore after each transaction.  Returning to the pool is able to reuse the connection which allows sqlite to to batched writes and minimize wear on storage.
- The volume of data we write has been reduced by 3-4x


A 1s commit interval makes a lot of sense because we can coalescing a lot of writes so we are always filling blocks, at 2s its a bit better but beyond that its diminishing returns.

It seems the recommendation for preserve SD card lifetime was changed to 30s in this PR https://github.com/home-assistant/home-assistant.io/pull/15065 which is likely quite overkill

Reply:
I think we can improve this for systems that have opted for a high commit interval, and as a bonus it will also improve the situation systems that where there is no pending work to be committed (we can start streaming sooner):

Create a new `async_synchronize` function on the recorder which does the following:

- ask the recorder if there is any pending commits, and if not return right away (probably will be the case on most systems)
- queue a commit task and wait for it to complete with a reasonable timeout in case there is something really broken with their system.

The amount of times we will have to commit a bit early should be infrequent. The amount of additional blocks written per day will likely pale in comparison to the delayed saves from serialized data such as the registries.

Reply:
Need to sleep on this one.  Something for a followup PR.

Reply:
Looks like the longest we can actually go without a commit is 5m since statistics runs will always force one

Reply:
Wasn't anywhere as large as I has expected so fixed in 05a7285d7c29e044372ca5b32118457646716bb6